### PR TITLE
Add vcs backend support

### PR DESCRIPTION
**Pull request by chick**
- Create new VcsBackendAnnotation
- Create support utilities for Vcs building tools
  - CopyVpiFiles move vpi and header files into place
  - GenVcsVerilogHarness  Compile stuff
  - VcsAnnotations control behavior of the Vcs backend
- Create VcsBackend, a trivial sub-classing of VerilatorBackend
- Create VcsExecutive
  - constructs VcsBackEnd requirements calls the constructor
- Create SimpleVcsTests
- Added String to Uniquify Verilator and Vcs Tests
- Test are ignored if VCS not present

### CODE REVIEW

**Comment by chick** [comment]
I am trying to find a setup where I can test this, will keep DO NOT MERGE label until then
But code should be reviewable now

File: src/main/scala/chisel3/tester/legacy/backends/vcs/CopyVpiFiles.scala [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    val vpiCppFilePath = Paths.get(destinationDirPath + "/vpi.cpp")
+    val vpiTabFilePath = Paths.get(destinationDirPath + "/vpi.tab")
+    try {
+      Files.createFile(simApiHFilePath)
+      Files.createFile(vpiHFilePath)
+      Files.createFile(vpiCppFilePath)
+      Files.createFile(vpiTabFilePath)
+    } catch {
+      case _: FileAlreadyExistsException =>
+        System.out.format("")
```
Comment (ducky64) [inline src/main/scala/chisel3/tester/legacy/backends/vcs/CopyVpiFiles.scala]: wat

File: src/test/scala/chisel3/experimental/tests/VerilatorBasicTests.scala [hunk 2]
```diff
@@ -8,7 +8,7 @@ import chisel3.tester.internal.VerilatorBackendAnnotation
 import org.scalatest._
 
 class VerilatorBasicTests extends FlatSpec with ChiselScalatestTester with Matchers {
-  behavior of "Testers2"
+  behavior of "Testers2_Verilator"
```
Comment (ducky64) [inline src/test/scala/chisel3/experimental/tests/VerilatorBasicTests.scala]: Maybe cleaner as "Testers2 with Verilator"? I don't think you need an underscore in quotes?

File: src/test/scala/chisel3/experimental/tests/VcsBasicTests.scala [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+import chisel3._
+import chisel3.tester._
+import chisel3.tester.experimental.TestOptionBuilder._
+import chisel3.tester.internal.VcsBackendAnnotation
+import chisel3.tests.{PassthroughModule, StaticModule}
+import org.scalatest._
+
+class VcsBasicTests extends FlatSpec with ChiselScalatestTester with Matchers {
+  behavior of "Testers2_Vcs"
```
Comment (ducky64) [inline src/test/scala/chisel3/experimental/tests/VcsBasicTests.scala]: ditto, "testesr2 with vcs"?

File: src/test/scala/chisel3/experimental/tests/VcsBasicTests.scala [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+  it should "test static circuits" in {
+    assume(firrtl.FileUtils.isVCSAvailable)
+
+    test(new StaticModule(42.U)).withAnnotations(annos) { c =>
+      c.out.expect(42.U)
+    }
+  }
+
+  it should "fail on poking outputs" in {
```
Comment (ducky64) [inline src/test/scala/chisel3/experimental/tests/VcsBasicTests.scala]: This seems more like a testers2 infrastructure exception, not sure if it belongs in a VCS test?

File: src/test/scala/chisel3/experimental/tests/VcsBasicTests.scala [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    assume(firrtl.FileUtils.isVCSAvailable)
+
+    assertThrows[exceptions.TestFailedException] {
+      test(new StaticModule(42.U)).withAnnotations(annos) { c =>
+        c.out.expect(0.U)
+      }
+    }
+  }
+
+  it should "fail with user-defined message" in {
```
Comment (ducky64) [inline src/test/scala/chisel3/experimental/tests/VcsBasicTests.scala]: Ditto, is this VCS-specific? I don't think so, or at least the control paths are redundant with the fail-on-expect simple case?

**Comment by sequencer** [comment]
Hello @chick. what’s the currently progress on vcs porting to tester2. 
I’m writing a new unittest framework for rocketchip with tester2 as backend. 
I think it would be really great to see it on mainstream.

**Comment by chick** [comment]
This should not be merged
It is superseded by PR #143

**Pull request closed without merging**